processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java

```diff
@@ -450,12 +452,173 @@ public Object apply(Cursor cursor)
+                      )
+                  );
+                }
+                catch (Exception ex) {
+                  throw new RuntimeException(ex);
+                }
+
+                DimensionSelector dimSelector1C = cursor.makeDimensionSelector(
+                    new DefaultDimensionSpec(
+                        "billy",
+                        "billy"
+                    )
+                );
+
+                DimensionSelector dimSelector2D = cursor.makeDimensionSelector(
+                    new DefaultDimensionSpec(
+                        "billy2",
+                        "billy2"
+                    )
+                );
+                //index gets more rows at this point, while other thread is iterating over the cursor
+                try {
+                  index.add(
+                      new MapBasedInputRow(
+                          timestamp,
+                          Lists.newArrayList("billy"),
+                          ImmutableMap.<String, Object>of("billy", "v3")
+                      )
+                  );
+                  index.add(
+                      new MapBasedInputRow(
+                          timestamp,
+                          Lists.newArrayList("billy3"),
+                          ImmutableMap.<String, Object>of("billy3", "")
+                      )
+                  );
+                }
+                catch (Exception ex) {
+                  throw new RuntimeException(ex);
+                }
+
+                DimensionSelector dimSelector3E = cursor.makeDimensionSelector(
+                    new DefaultDimensionSpec(
+                        "billy3",
+                        "billy3"
+                    )
+                );
+
+                CountDownLatch rowNumInCursor = new CountDownLatch(1);
+                // and then, cursoring continues in the other thread
+                while (!cursor.isDone()) {
+                  IndexedInts rowA = dimSelector1A.getRow();
+                  for (int i : rowA) {
+                    Assert.assertTrue(i < cardinalityA);
+                  }
+                  IndexedInts rowB = dimSelector1B.getRow();
+                  for (int i : rowB) {
+                    Assert.assertTrue(i < cardinalityA);
+                  }
+                  IndexedInts rowC = dimSelector1C.getRow();
+                  for (int i : rowC) {
+                    Assert.assertTrue(i < cardinalityA);
+                  }
+                  IndexedInts rowD = dimSelector2D.getRow();
+                  // no null id, so should get empty dims array
+                  Assert.assertEquals(0, rowD.size());
+                  IndexedInts rowE = dimSelector3E.getRow();
+                  Assert.assertEquals(1, rowE.size());
+                  // the null id
+                  Assert.assertEquals(0, rowE.get(0));
+                  cursor.advance();
+                  rowNumInCursor.countDown();
```

Comment:
Please use AtomicInteger + increment for this. Otherwise, we can't properly measure the number of rows that get returned by the cursor. (countDown on a latch with value 0 is ignored)